crates/hyperion/src/simulation/inventory.rs

```diff
@@ -1191,6 +1208,74 @@ fn handle_drop_key(
     event_writer.write(event);
 }
 
+fn validate_slot_changes(
+    slot_changes: &[SlotChange],
+    inventories_mut: &[&mut ItemSlot],
+    cursor_item: &ItemStack,
+    player_only: bool,
+) -> bool {
+    // If cursor is empty, no slot changes should be valid
+    if cursor_item.is_empty() {
+        return false;
+    }
+
+    for slot_change in slot_changes {
+        // Validate slot index bounds
+        let Ok(slot_idx) = usize::try_from(slot_change.idx) else {
+            return false;
+        };
+
+        // Check if slot exists in our inventory
+        if slot_idx >= inventories_mut.len() {
+            return false;
+        }
+
+        let slot = &inventories_mut[slot_idx];
+
+        // Skip readonly slots
+        if slot.readonly {
+            return false;
+        }
+
+        // For player-only inventories, validate armor slot restrictions
+        if player_only && (5..=8).contains(&slot_idx) {
+            let is_valid = match slot_idx {
+                5 => cursor_item.item.is_helmet(),
+                6 => cursor_item.item.is_chestplate(),
+                7 => cursor_item.item.is_leggings(),
+                8 => cursor_item.item.is_boots(),
+                _ => true,
+            };
+            if !is_valid {
+                return false;
+            }
+        }
+
+        // Validate that the slot is either empty or contains the same item type
+        // This prevents creating items out of thin air
+        if !slot.stack.is_empty() {
+            // The slot should either be empty or contain the same item type as cursor
+            if slot.stack.item != cursor_item.item || slot.stack.nbt != cursor_item.nbt {
+                return false;
+            }
+
+            // For non-empty slots, they should have space available
+            if slot.stack.count >= slot.stack.item.max_stack() {
+                return false;
+            }
+        }
+    }
+
+    // Validate that the total number of items we're trying to distribute
+    // doesn't exceed what's in the cursor
+    let total_distributed_items = i32::try_from(slot_changes.len()).unwrap();
+    if total_distributed_items > i32::from(cursor_item.count) {
```

Comment:
I'd recommend changing this to `==` so players can't destroy their own items
```suggestion
    if total_distributed_items == i32::from(cursor_item.count) {
```

Reply:
If you have 32 diamonds in cursor and drag over 6 slots, you'd distribute ~5 per slot with 2 remaining. With ==, this would be rejected because 6 != 32